splitio/producer/storage/segment_wrapper.go

```diff
@@ -0,0 +1,141 @@
+package storage
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/splitio/go-split-commons/v4/storage"
+	"github.com/splitio/go-toolkit/v5/datastructures/set"
+	"github.com/splitio/go-toolkit/v5/logging"
+)
+
+// ObservableSegmentStorage builds on top of the SegmentStorage interface adding some observability methods
+type ObservableSegmentStorage interface {
+	storage.SegmentStorage
+	NamesAndCount() map[string]int
+}
+
+// ObservableSegmentStorageImpl is an implementation of the ObservableSegmentStorage interface
+type ObservableSegmentStorageImpl struct {
+	storage.SegmentStorage
+	counter *activeSegmentTracker
+	logger  logging.LoggerInterface
+}
+
+// NewObservableSegmentStorage constructs and observable segment storage
+func NewObservableSegmentStorage(
+	logger logging.LoggerInterface,
+	splitStorage storage.SplitStorage,
+	toWrap storage.SegmentStorage,
+) *ObservableSegmentStorageImpl {
+
+	if _, ok := toWrap.(supportsUpdateWithSummary); !ok {
+		logger.Warning("supplied segment storage doesn't report added/removed counts, this may introduce inconcistencies in observability endpoints")
+	}
+
+	segmentNames := splitStorage.SegmentNames()
+	tracker := newActiveSegmentTracker(segmentNames.Size() + 1)
+
+	if extSS, ok := toWrap.(supportsCardinality); ok {
+		segmentNames.Each(func(i interface{}) bool {
+			strName, ok := i.(string)
+			if !ok {
+				logger.Warning(fmt.Sprintf("non-string segment name fetched: '%+v'//'%T'. This is a bug, please report it.", i, i))
+				return true
+			}
+
+			count, err := extSS.Size(strName)
+			if err != nil {
+				logger.Warning(fmt.Sprintf("failed to get size for segment %s. This may introduce inconsistencies in observability endpoints", strName))
+			}
+
+			tracker.update(strName, count, 0)
+			return true
+		})
+	} else {
+		logger.Warning("storage does not support querying segment size. This may introduce inconsistencies in observability endpoints")
+	}
+
+	return &ObservableSegmentStorageImpl{
+		SegmentStorage: toWrap,
+		counter:        tracker,
+		logger:         logger,
+	}
+}
+
+// Update updates the local segment cache and forwards the call to he underlying storage
+func (s *ObservableSegmentStorageImpl) Update(name string, toAdd *set.ThreadUnsafeSet, toRemove *set.ThreadUnsafeSet, changeNumber int64) error {
+	var added, removed int
+	if rs, ok := s.SegmentStorage.(supportsUpdateWithSummary); ok {
+		var err error
+		added, removed, err = rs.UpdateWithSummary(name, toAdd, toRemove, changeNumber)
+		if err != nil {
+			// TODO(mredolatti): log!
+		}
+	} else {
+		// TODO(mredolatti): is this worth logging? are we just going to annoy people?
```

Comment:
I don't think it's worth it - you already logged when instantiating the class ("this won't support certain observability features"). There won't be anything changing on runtime (I believe) to change that, so might as well just consider it already reported.
Related - see my other note about not even allowing this to happen.

Reply:
+1 to that